vllm/model_executor/layers/quantization/modelopt.py

```diff
@@ -1089,23 +1269,29 @@ def apply(
         expert_load_view: Optional[torch.Tensor] = None,
         logical_to_physical_map: Optional[torch.Tensor] = None,
         logical_replica_count: Optional[torch.Tensor] = None,
+        routed_scaling_factor: Optional[float] = 1.0,
     ):
         if enable_eplb:
             raise NotImplementedError(
                 "EPLB not supported for `ModelOptNvFp4FusedMoE` yet.")
         assert activation == "silu", "Only SiLU activation is supported."
 
-        topk_weights, topk_ids = FusedMoE.select_experts(
-            hidden_states=x,
-            router_logits=router_logits,
-            use_grouped_topk=use_grouped_topk,
-            top_k=top_k,
-            renormalize=renormalize,
-            topk_group=topk_group,
-            num_expert_group=num_expert_group,
-            custom_routing_function=custom_routing_function,
-            scoring_func=scoring_func,
-            e_score_correction_bias=e_score_correction_bias)
+        skip_select_experts = False
+        if self.fused_experts is not None and \
+            self.flashinfer_moe_backend == "TensorRT-LLM":
+            skip_select_experts = True
+        if not skip_select_experts:
```

Comment:
You could move the block for flashinfer.fused_moe.trtllm_fp4_block_scale_moe before FusedMoE.select_experts and return here early